### ISSUE TITLE
How does static metaprogramming handle incomplete code for auto-complete?

### ISSUE THREAD

**Issue by munificent** [dart-lang/language#1628]
We discussed this in a meeting, but I wanted to capture it in an issue so we don't forget.

Currently, analyzer is happy to analyze incomplete or incorrect programs in order to make auto-complete work while the user is in the middle of typing code. For example:

<img width="543" alt="Screen Shot 2021-05-11 at 3 23 55 PM" src="https://user-images.githubusercontent.com/46275/117892096-f154a300-b26c-11eb-9d80-1cf5e9bea219.png">

Here, even though the class is syntactically incorrect, it can still type check the class enough to show `bar()` as an auto-complete option. If something like macros are applied to that class, though, this can break down. If the analyzer/compiler only applies macros when the code is in a valid state, then auto-complete will stop working while they are typing in a class that uses macros. If analysis can apply macros to incomplete code, then we need to define how that works or what that means.

One idea is that the language basically ignores the issue. It only specifies the behavior of applying macros to correct programs. When the user is in the middle of typing code, analyzer takes their current incomplete program and discards whatever incomplete stuff it needs to get to a valid source file. Then it runs the macros and type checks that. Maybe something like that would work.

Whatever we do, it's critical that static metaprogramming doesn't degrade the IDE user experience.

@bwilkerson @jakemac53

**Comment by bwilkerson** [comment]
@devoncarew, @scheglov and I recently had the first in a series of meetings about some of the UX issues around static metaprogramming and I'm in the process of writing up a doc with what we had thought about. Our intent is to share it with you as soon as it's reasonably coherent as a way of furthering discussions like this. Code completion is one of the topic areas, and I'll be sure to add thoughts about incomplete code.

**Comment by stereotype441** [comment]
Definitely looking forward to @bwilkerson's doc.  As someone who's been hacking on the analyzer for 7 years, I have a bunch of thoughts on this topic too.  I'll try to give my personal perspective here briefly; I'm happy to follow up off-line if that would be useful.

The issue IMHO is better characterized as how to handle *code with static errors* rather than *incomplete* code.  (Some incomplete code is free of static errors and that code can be analyzed normally; some code with static errors isn't incomplete but actually has extra junk, or stuff that's simply wrong (like code that fails to type check) and we need to handle that too).  So really what we are talking about here is error recovery.

I like the simplicity of Bob's suggestion that we discard code until we reach a valid program, but I think in order to reduce the amount of discarding we need to do, we will need some escape hatches to allow invalid code to be represented.  As a trivial example, if a method's return type is misspelled, we shouldn't just discard that method from the model; the code generator should still be able to see that the method exists.  But if it queries its return type, it should either get `dynamic`, or a special "invalid type" with some well-defined behaviors.  Personally I favor the "invalid type" approach because it makes it easier to avoid cascading errors, but that may be more implementation effort because currently the analyzer uses `dynamic`.

Similarly, if we allow code generators to peek at the expressions and statements inside a method declaration, it would be nice to have "invalid statement" and "invalid expression" nodes so that we can show the code generator as much of the method body as possible without forcing it to deal directly with the invalid code.

A particularly tricky area is what to do with loops in class hierarchies.  The language spec forbids the class hierarchy from looping in order to avoid non-termination of algorithms like type inference, subtype checking, etc.  The analyzer's implementation of these algorithms has been carefully written to avoid non-termination even in the presence of class hierarchy loops.  I don't think we can expect writers of code generators to exercise the same care, and if we're going to be running user-supplied code generators during analysis, it's really important to avoid non-termination issues.  So I think we should ensure that when their code generator queries the code structure, it sees a class hierarchy without loops.  We *could* handle this by discarding "extends" and "implements" clauses until no loops remain, but I think this would lead to a poor user experience because code generators often have expectations about the class structure of the code they're processing, and will issue errors if they don't see the class structure they expect.  It would be really confusing if our efforts to "fix" class hierarchy loops caused code generators to output nonsense error messages.  I think a better approach here is the "invalid type" approach: if there's a loop in the class hierarchy, the analyzer should synthetically change one of the "extends" or "implements" clasuses to e.g. "extends <invalid-type>"; that way either the code generator can choose to suppress its error message, or even if it doesn't and the error message makes it to the user, the name "invalid-type" should provide the user with better clues about what went wrong.

Side question: in order to fix loops in the class hierarchy, the analyzer will need to make arbitrary choices about which edge to cut.  How consistent should its choices be?  IMHO it would be best for it to be deterministic and independent of the order of queries from the code generator.

One final note: thus far, we have left the design of error recovery pretty much entirely up to the analyzer team, and kept any discussion of error recovery out of the spec.  I think that's a really good general principle because it allows the analyzer team more creative freedom and lets them pivot faster when they learn things about how error recovery works in real-world scenarios.  But I think it would be good to classify our errors into some broad categories and specify what categories of invalid code (if any) user-provided code should be expected to deal with.  I think it should be something like: user-provided code generators won't ever see syntax errors, class hierarchy loops, or self-referential typedefs.  But they may see special "invalid type" nodes.  They may also see: unimplemented abstract methods, type mismatches between a class method and the method it overrides, and naming conflicts (e.g. classes with duplicate names).  Also, when looking inside method bodies, they may see "invalid expression" and "invalid statement" nodes, unterminated switch cases, mismatched return types, use of uninitialized variables, and failure to return a value from a method with a non-nullable return type.

(That's just my first impression though.  I'm fairly open to discussion, especially on that last sentence about method bodies)

**Comment by bwilkerson** [comment]
I agree, this is largely a question of recovery support.

> If the analyzer/compiler only applies macros when the code is in a valid state, then auto-complete will stop working while they are typing in a class that uses macros.

Completion isn't going to stop working. The analysis server will degrade gracefully, and the doc will outline our current thinking on what we expect that to mean in practice.

> If analysis can apply macros to incomplete code, then we need to define how that works or what that means.

We certainly need to define what the APIs used by macros do in the face of incomplete code. What happens when the macro asks about the superclass of a class and the name in the extends clause doesn't exist? What happens when, as Paul mentioned, there's a circularity in the class hierarchy? There are lots of interesting questions in this space. To be clear, those questions are outside the scope of the doc we're putting together.

And as has been said before, we need to define the conditions under which macros should not be run, if any, such as syntactic errors. It isn't clear to me, though, that the answer needs to be the same for all tools. For everything other than macros, a command-line compiler can typically choose a subset of conditions from which to recover while IDE support needs to recover from every possible error. I think the situation is analogous with macros. If a command-line compiler detects and exception in a macro it might choose to not do any recovery, whereas the IDE support will need to recover from that exception in some way.

> One idea is that the language basically ignores the issue. It only specifies the behavior of applying macros to correct programs.

I think that's almost reasonable. The language doesn't, and shouldn't, specify how tools recover from other errors, so it isn't necessary for the language to specify how tools recover from errors in macros.

The one possible exception, as I mentioned before, is the APIs used by macros to query the user written code. I don't know how you're planning on handling those APIs. Will the language specify the APIs to be used, or merely specify the syntax and semantics of macros and leave the API unspecified? The APIs need to specify their behavior in the face of errors. For example, do they throw exceptions, return some kind of error object, or leave the behavior undefined?

Paul's comment is largely addressing this question, and I'd be happy to weigh in on specific ideas, but I'd like to know first whether we're going to try require the same recovery approach everywhere or whether individual tools have the freedom to vary their approach. It might significantly change my answer in some cases.

> When the user is in the middle of typing code, analyzer takes their current incomplete program and discards whatever incomplete stuff it needs to get to a valid source file. Then it runs the macros and type checks that. Maybe something like that would work.

That's an interesting recovery strategy, but it's not consistent with the approach used by the analyzer for other kinds of errors. I'm not saying that it has to be consistent, just making an observation. I think it would be informative to try to write a couple of macros as a concrete way of thinking about which recovery approach would provide the best UX for both end users and for macro authors.

**Comment by munificent** [comment]
This is a great discussion. Trying to synthesize your comments, how does something like this sound:

* Tools may run macros even on code that contains some kind of errors. It's entirely up to tools to choose whether and in which circumstances that they do. The language doesn't specify the boundary.
* But, if the code isn't valid, it will never end up being run. The macros are only invoked in order to provide a better IDE experience.
* The API that macros use to work with code has a notion of erroneous pieces of code. Introspecting over code in a macro may yield "invalid type", "invalid statement", etc. Macro authors can choose to handle those gracefully if they choose in order to provide a better experience when the macro is invoked on invalid code. If they don't the macro may fail, which essentially means it generates some cascaded compile errors. It's up to tools to decide if they want to hide those or not.

 In other words tools *may* eagerly run macros on ill-defined code. The API gives macro authors affordances to handle that gracefully if they so choose. But the resulting code never gets actually executed, so the language doesn't have to get in the business of specifying its behavior.

Does that seem like a reasonable set of trade-offs?

**Comment by tatumizer** [comment]
> If the analyzer/compiler only applies macros when the code is in a valid state, then auto-complete will stop working while they are typing in a class that uses macros

Why will it stop working? 
Consider a macro that generates a data class. Let's assume the input for a macro is declared as "template class", which is easily recognized by the analyzer as a template (as opposed to a normal code). 
```dart
@dataClass
template class Foo {
  int x;
  String y;
  // etc.
}
```
Further, suppose we already have a state with a valid template. The macro has been called already, and the resulting code is known. What happens when we move the cursor into the definition of the template and start editing?
The straightforward answer is: nothing. The macro doesn't get called when the code of the template is incomplete. But the previous results of a macro are not discarded - at least until the cursor leaves the template block. So all suggestions work as if nothing had happened. If the user exits the block leaving the template in a bad state - I'm not sure even in this case the results of prior macro invocation should be invalidated (thus causing an avalanche of other errors), but maybe.  

However, if the macro gets called with incomplete code (as proposed) and has to decide how to handle the errors "gracefully" - it's painful even to think of how the macro has to be coded to behave predictably. The only possibility I see for the macro is to memorize the input and output of the last correct invocation and somehow figure out what has changed and how this affects the result - but why bother? There's no guarantee that this behavior will be more intuitive than the straightforward algorithm outlined above.

**Comment by bwilkerson** [comment]
> Tools may run macros even on code that contains some kind of errors. It's entirely up to tools to choose whether and in which circumstances that they do. The language doesn't specify the boundary.

Agreed.

> The API that macros use to work with code has a notion of erroneous pieces of code. Introspecting over code in a macro may yield "invalid type", "invalid statement", etc.

> However, if the macro gets called with incomplete code (as proposed) and has to decide how to handle the errors "gracefully" - it's painful even to think of how the macro has to be coded to behave predictably.

The issue we're discussing is recovery, and so far we've been discussing _how_ to recover. But this raises the question of _who_ should be responsible for recovery. My preference would be to put the burden for recovery on the tools (analyzer in this case) but to allow (not require) macros to recover in cases where there's a clear use case for it. Doing so will ease the burden for macro authors and it won't increase the amount of work for the tools.

I don't have any clear use cases to offer at this point, so I'd start with assuming we'll have no support for macros to participate in recovery. I would, however, propose that we should run through some use cases for macros to think about whether there are times when a macro would want to control recovery in the face of invalid code. I'll also note that if we find a need after the first version has shipped we can always extend the API to handle this later.

Here's my proposal: the analyzer can speculatively run a macro and if the macro needs information that isn't available because of invalid code it can stop running the macro and respond as if it had decided not to run the macro in the first place. That approach
- provides a clear definition of the circumstances under which macros will be run, at least as far as dealing with invalid code is concerned,
- allows the analyzer to have the maximum amount of up-to-date information, and
- doesn't require macro authors to think about recovery.

I suspect that it will also produce an optimal UX for users.

So, no, I don't think the API needs to have a representation for invalid code (unless we find a compelling use case).

> But the previous results of a macro are not discarded ...

Yep, that's part of the proposal I'm writing up.

**Comment by jakemac53** [comment]
> I'll also note that if we find a need after the first version has shipped we can always extend the API to handle this later.

Not necessarily, it depends a lot on the API we ship. We would need to at least design the api of this feature first to ensure that we could ship it as a non-breaking feature. 

> I don't have any clear use cases to offer at this point, so I'd start with assuming we'll have no support for macros to participate in recovery.

There is actually one case which I think might fall into this same bucket as what we are discussing here. The case is where a macro introspects on an api, but the api relies on type inference to be filled in, and the macro itself needs to be ran in order for that type inference to work. Consider this class:

```dart
class Thing {
  final name = someMacroGeneratedField;
}
```

If the macro generating `someMacroGeneratedField` introspects on `Thing` and asks for the type of `name`, what does it get? 

The issue is actually a lot more subtle/interesting than even this contrived example, because even if `someMacroGeneratedField` can be resolved to some identifier, some other macro could run and change the identifier that should have been resolved to by shadowing it.

This is a general problem that we are working through right now but it seems likely that macros will need to at least have to deal with introspecting on apis where the type is _not yet known_, and never will be whenever that macro runs, because it hasn't been confidently inferred yet.

My assumption at this point is some macros would want to fail and require a type on the left hand side (although this is not ideal, I don't see a good workaround at this time). Other macros might be ok with just treating it as dynamic. I think we need to give them some special notion of an "unknown type" to choose how to handle this.

**Comment by bwilkerson** [comment]
> ... where the type is not yet known, and never will be whenever that macro runs, because it hasn't been confidently inferred yet.

I think there's an interesting requirement implied by that statement: while applying the macro M, the introspection API must reveal a state that's consistent with the state the code would be in if the macros were being run for the first time. In other words, if the analyzer caches any macro results and then has to re-apply M, the introspection API can't reveal any code in the cache that would have been generated after M was run the first time.

@devoncarew @scheglov

**Comment by scheglov** [comment]
The requirement that every time is like the first time looks reasonable to me.

**Comment by davidmorgan** [comment]
The current status is that macros run on incomplete code, even syntactically incorrect code, and can give sufficient results so that autocomplete can work.

Please reopen if there is any need to change / discuss changes. Thanks :)
